File: 1696-jump-game-vi/1696-jump-game-vi.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        
        

        int N = nums.size() ; 
        vector<int> dp(N) ; 
        dp[0]= nums[0] ; 
        deque<int> q ; 
        q.push_back(0); 

        for( int i=1; i < N ; i++){

            while( !q.empty() && q.front() < i-k){
                q.pop_front() ; } 

            dp[i] = dp[q.front()] + nums[i] ; 

            while( !q.empty() &&  dp[i] >= dp[q.back()] ){
                q.pop_back() ; } 

            q.push_back(i) ;  
        }
        return dp[N-1] ; 


        
        
        
        
    }
};

File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        
        import heapq as hq 
        
        dp = [0]*(len(nums))
        dp[0] = nums[0]
        heap = [(-nums[0],0)]
        for i in range(1, len(nums)):
            
            while heap[0][1] < i-k: 
                hq.heappop(heap)
            
            dp[i] = nums[i] + dp[heap[0][1]]
            hq.heappush(heap ,( -dp[i] , i)) 
        return dp[-1]
        